### ISSUE TITLE
Unable to properly infer numeric keys in TypeScript type with mixed key types

### ISSUE THREAD

**Issue by aronmal** [adaltas/node-http-status#48]
**Describe the bug**

When utilizing the `http-status` package, there is a bug related to inferring numeric keys in TypeScript when working with a mixed type declaration within the `HttpStatus` interface in [index.d.ts](https://github.com/adaltas/node-http-status/blame/71a991bc4efa131985e97e35248f41bb59c62e6b/lib/index.d.ts).

**To Reproduce**

To reproduce the issue, consider the following TypeScript code snippet:

```typescript
import httpStatus from "http-status"; // object itself
import type { HttpStatus } from "http-status"; // interface

// example for comparison
const testStatus = {
  100: "",
  "100_NAME": "",
  "100_MESSAGE": "",
  CONTINUE: 100,
  101: "",
  "101_NAME": "",
  "101_MESSAGE": "",
  SWITCHING_PROTOCOLS: 101,
  102: "",
  "102_NAME": "",
  "102_MESSAGE": "",
  PROCESSING: 102,
  103: "",
  "103_NAME": "",
  "103_MESSAGE": "",
  EARLY_HINTS: 103,
};

// Define a custom type to extract numeric keys
type NumberKeys<T> = {
  [K in keyof T]: T[K] extends number ? (string extends K ? never : K) : never;
}[keyof T];

// Define a function with a parameter that should only accept numeric keys
export function Response(code: NumberKeys<HttpStatus>) {
  const thisCode = httpStatus[code];
  // Rest of your function logic
}
```

In this case, the types do not work correctly and autocompletion does show any possible string in the ide 
![image](https://github.com/adaltas/node-http-status/assets/57275207/e77d7d01-6de1-4929-aa21-bf3e9e7b30bc)

When changing the type input to the example object:

```diff
- export function Response(code: NumberKeys<HttpStatus>) {
+ export function Response(code: NumberKeys<typeof test>) {
    const thisCode = httpStatus[code];
    // Rest of your function logic
  }
```

it then works and shows the correct types

![image](https://github.com/adaltas/node-http-status/assets/57275207/b3339ee9-8cf1-448a-9bf9-7ad2e1a49fff)

Additionally, the unnecessary declaration `readonly [key: number]: string | undefined;` in the `HttpStatus` interface contributes to the issue and might be impacting TypeScript type inference.

**Additional context**

The issue seems to be associated with the mixed type declaration within the `HttpStatus` interface  in [index.d.ts](https://github.com/adaltas/node-http-status/blame/71a991bc4efa131985e97e35248f41bb59c62e6b/lib/index.d.ts), specifically:

```typescript
export = httpStatus;

declare const httpStatus: httpStatus.HttpStatus;

declare namespace httpStatus {
  interface HttpStatus {
    // not important for this issue, but I still see no sense in this either
    readonly [key: number]: string | undefined;

    // this causes the issue
    readonly [key: string]:
    | string
    | number
    | HttpStatusClasses
    | HttpStatusExtra
    | undefined;
```

This declaration allows for both string and number keys, leading to unexpected behavior in TypeScript type inference. The bug impacts TypeScript type inference in projects utilizing the `http-status` package, particularly when trying to enforce strict typing for numeric keys. The unnecessary `readonly [key: number]` declaration are causing the issue and should be considered for removal.

**Comment by aronmal** [comment]
Is there a reason for the `readonly [key: number]`? If not I can create a pull request, but I wanted to ask first if it has an origin.

**Comment by wdavidw** [comment]
I don't think so, or I don't have a memory for it. Please propose your change if you feel confident. I personally don't use TS.

**Comment by wdavidw** [comment]
Also, add a unit test to support your claim.

**Comment by aronmal** [comment]
> Also, add a unit test to support your claim.

I've addressed the comment about adding a unit test. While there isn't a specific test to showcase the issue due to the nature of the problem (lack of strictness in types allowing more keys than intended), I did enhance the existing unit tests. 

During this process, an type error in the test file surfaced, but I also felt free to do a comprehensive rework of the typing. Previously, explicit type definitions were necessary to compensate for the lack of strictness. With the improvements, the need for these explicit types has been significantly reduced, and removing them has proven to be effective in resolving the issue as well.

If you have any further suggestions, let me know!
